### PR TITLE
Add backward for spdiags (cpu) and enable autograd

### PR DESCRIPTION
**Pull request by amjames**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79462
* #79228
* __->__ #78965

cc @nikitaved @pearu @cpuhrsch @bhosmer

### CODE REVIEW

**Review by cpuhrsch — APPROVED** [review]

File: aten/src/ATen/native/cpu/SparseFactories.cpp [hunk 1]
```diff
@@ -66,9 +66,30 @@ void _spdiags_kernel_cpu(
       });
 }
 
+void _spdiags_backward_kernel_cpu(TensorIterator& iter, Tensor& grad_in) {
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(
+      at::ScalarType::BFloat16,
+      at::ScalarType::Half,
+      at::ScalarType::Bool,
+      at::ScalarType::ComplexHalf,
```
Comment (pearu) [inline aten/src/ATen/native/cpu/SparseFactories.cpp:74]: No reason to support non-float/complex types:
```suggestion
  AT_DISPATCH_CASE_FLOATING_AND_COMPLEX_TYPES_AND3(
      at::ScalarType::BFloat16,
      at::ScalarType::Half,
      at::ScalarType::ComplexHalf,
```

**Review by pearu — CHANGES_REQUESTED** [review]
I have a few nits, overall looks good. Thanks, @amjames!

File: aten/src/ATen/native/sparse/SparseFactories.cpp [hunk 2]
```diff
@@ -91,5 +95,48 @@ Tensor spdiags(
...
+
+  auto n_diag = input_shape.size() == 2 ? input_shape[0] : 1;
+  auto n_col_in = input_shape.size() == 2 ? input_shape[1] : input_shape[0];
+  AT_ASSERT(grad_out.dim() == 2);
+  AT_ASSERT(offsets_1d.size(0) == n_diag);
+  auto grad_in_options = grad_out.options().layout(Layout::Strided);
+  // zeros since we are only going to se the non-zero elements
+  Tensor grad_in = at::zeros({n_diag, n_col_in}, grad_in_options);
+  auto grad_out_coo =
+      grad_out.layout() == Layout::Sparse ? grad_out : grad_out.to_sparse();
```
[the remark this answers is not in the document]
Comment (amjames) [reply aten/src/ATen/native/sparse/SparseFactories.cpp:112]: I am assuming grad_out will have the same sparse layout at the output of the forward pass, (if called `spdiags(..., layout=torch.sparse_csr)` for example). So the conversion to coo is needed.

I am not sure if `grad_out` may be uncoalesced if the layout is coo. The kernel function as implemented here would not be correct for that case, but it would be a simple modification to make it correct for that case, as we are already zero initializing `grad_in`.

File: aten/src/ATen/native/cpu/SparseFactories.cpp [hunk 3]
```diff
@@ -66,9 +66,30 @@ void _spdiags_kernel_cpu(
...
+      iter.dtype(),
+      "spdiags_backward_cpu",
+      [&] {
+        auto grad_in_accessor = grad_in.accessor<scalar_t, 2>();
+        cpu_kernel(
+            iter,
+            [&](scalar_t grad_out_value,
+                int64_t row_idx,
+                int64_t col_idx) -> scalar_t {
+              grad_in_accessor[row_idx][col_idx] = grad_out_value;
```
Comment (amjames) [inline aten/src/ATen/native/cpu/SparseFactories.cpp:84]: @pearu changing this line to `grad_in_accessor[row_idx][col_idx] += grad_out_value` should correctly handle the uncoalesced `grad_out` case?

File: test/test_sparse.py [hunk 4]
```diff
@@ -3566,12 +3566,13 @@ def test(sparse_dims, nnz, with_size, new_size):
     @dtypes(*all_types_and_complex_and(torch.bool))
     def test_sparse_spdiags(self, device, dtype):
 
-        make_diags = functools.partial(make_tensor, dtype=dtype, device=device)
+        requires_grad = dtype in (torch.double, torch.cdouble)
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply test/test_sparse.py:3569]: Yes, run `gradcheck` only on doubles always!

File: aten/src/ATen/native/sparse/SparseFactories.cpp [hunk 5]
```diff
@@ -91,5 +95,49 @@ Tensor spdiags(
...
+  auto n_diag = input_shape.size() == 2 ? input_shape[0] : 1;
+  auto n_col_in = input_shape.size() == 2 ? input_shape[1] : input_shape[0];
+  AT_ASSERT(grad_out.dim() == 2);
+  AT_ASSERT(offsets_1d.size(0) == n_diag);
+  auto grad_in_options = grad_out.options().layout(Layout::Strided);
+  // zeros since we are only going to se the non-zero elements
+  Tensor grad_in = at::zeros({n_diag, n_col_in}, grad_in_options);
+  auto grad_out_coo = grad_out.layout() == Layout::Sparse
+      ? grad_out.coalesce()
+      : grad_out.to_sparse();
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseFactories.cpp:113]: Could `grad_out` be a non-sparse Tensor?

File: aten/src/ATen/native/sparse/SparseFactories.cpp [hunk 6]
```diff
@@ -91,5 +95,49 @@ Tensor spdiags(
...
+  auto grad_out_indices = sparse::get_sparse_impl(grad_out_coo)->indices_;
+  auto grad_out_row_indices = grad_out_indices[0];
+  auto grad_out_col_indices = grad_out_indices[1];
+
+  // Precompute input row indices for each nnz
+  auto row_in_indices =
+      offsets_1d
+          .eq(grad_out_col_indices.sub(grad_out_row_indices).reshape({-1, 1}))
+          .nonzero()
+          .permute({1, 0})[-1];
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseFactories.cpp:125]: Why do you make it a 2D tensor again?

File: test/test_sparse.py [hunk 7]
```diff
@@ -3566,12 +3566,13 @@ def test(sparse_dims, nnz, with_size, new_size):
...
+        make_diags = functools.partial(make_tensor, dtype=dtype, device=device, requires_grad=requires_grad)
         make_offsets = functools.partial(torch.tensor, dtype=torch.long, device=device)
 
         if TEST_SCIPY:
             def reference(diags, offsets, shape):
-                return scipy.sparse.spdiags(diags, offsets, *shape).toarray()
+                return scipy.sparse.spdiags(diags.detach().cpu(), offsets.cpu(), *shape).toarray()
```
Comment (nikitaved) [inline test/test_sparse.py:3575]: Just curious, is `detach` necessary here?

File: tools/autograd/derivatives.yaml [hunk 8]
```diff
@@ -2956,3 +2956,13 @@
 
 - name: special_spherical_bessel_j0(Tensor x) -> Tensor
   x: non_differentiable
+
+- name: _spdiags(Tensor diagonals, Tensor offsets, int[] shape, Layout? layout=None) -> Tensor
+  diagonals: _spdiags_backward(grad, offsets, diagonals.sizes())
+  offsets: non_differentiable
+  result: auto_linear
+
```
Comment (nikitaved) [inline tools/autograd/derivatives.yaml:2964]: Is forward AD being tested somewhere?

File: aten/src/ATen/native/sparse/SparseFactories.cpp (hunk 6 above)
Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseFactories.cpp:125]: I think this could be simplified to `.eq(grad_out_col_indices - grad_out_row_indices).nonzero().squeeze(-1)`.

File: aten/src/ATen/native/sparse/SparseFactories.cpp [hunk 9]
```diff
@@ -91,5 +95,49 @@ Tensor spdiags(
...
+
+  // dummy output required by cpu/gpu_kernel
+  auto dummy = at::empty({1}, grad_in.options());
+  auto iter = TensorIteratorConfig()
+                  .check_all_same_dtype(false)
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseFactories.cpp:130]: I think this is set to `false` by default, no? Maybe you meant `set_check_mem_overlap(false)`? Or if this is needed, let's have them both!

File: test/test_sparse.py (hunk 7 above)
Comment (amjames) [reply test/test_sparse.py:3575]: When passing the tensor to SciPy there will be a `.numpy()` conversion done. If `diags` has `requires_grad=True` this causes a runtime error: `Can't call numpy() on a Tensor that requires grad. Use tensor.detach().numpy() instead`.

File: aten/src/ATen/native/sparse/SparseFactories.cpp (hunk 5 above)
Comment (amjames) [reply aten/src/ATen/native/sparse/SparseFactories.cpp:113]: `grad_out` will always be a sparse tensor, but may have `sparse_csr`, or `sparse_csc` layout. It should never be a dense tensor though as `spdiags` will always output a sparse tensor.

File: aten/src/ATen/native/sparse/SparseFactories.cpp (hunk 6 above)
Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseFactories.cpp:125]: `grad_out_col_indices` and `grad_out_row_indices` are 1D tensors, so `nonzero` will give you the `(nnz, 1)` shape.

Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseFactories.cpp:125]: `grad_out_col_indices` and `grad_out_row_indices` are 1D tensors already, so `nonzero` is going to give you the `(nnz, 1)` shape.

Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseFactories.cpp:125]: Too tired... Sorry for the noise. Right, so, `grad_out_col_indices` and `grad_out_row_indices` are 1D already, so `nonzero` will give you a `(nnz, 1)` shape.

Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseFactories.cpp:125]: Right, sorry, missed the order. Are these the same offsets from the forward?

Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseFactories.cpp:125]: Nevermind, offsets_1d has a different shape from the indices.

File: tools/autograd/derivatives.yaml (hunk 8 above)
Comment (amjames) [reply tools/autograd/derivatives.yaml:2964]: No that is an oversight on my part. I think the best path to integrating forward AD tests is simply to port the whole test suite to `OpInfos`.

**Comment by amjames** [comment]
This is going to be on hold until I finish some format interconversions. Once we have these in the tests can port to `OpInfos` (which will ensure forward AD is tested among other things). 

I will ping for next round of reviews once that is ready.
